bin/que

```diff
@@ -84,24 +94,51 @@ rescue NameError
   exit 1
 end
 
-queue_name          = options.queue_name     || ENV['QUE_QUEUE'] || Que::Worker::DEFAULT_QUEUE
-wake_interval       = options.wake_interval  || ENV['QUE_WAKE_INTERVAL']&.to_f
-metrics_labels      = options.metrics_labels || {}
+queue_name          = options.queue_name         || ENV['QUE_QUEUE'] || Que::Worker::DEFAULT_QUEUE
+wake_interval       = options.wake_interval      || ENV['QUE_WAKE_INTERVAL']&.to_f
+metrics_labels      = options.metrics_labels     || {}
 priority_threshold  = options.priority_threshold || ENV['QUE_PRIORITY_THRESHOLD'].to_i
-cursor_expiry       = options.cursor_expiry  || 0
+cursor_expiry       = options.cursor_expiry      || 0
+timeout             = options.timeout            || 0
+timeout_message     = options.timeout_message    || ""
 
 Que.logger.info(msg: 'Starting worker', event: 'que.worker.start')
 
 worker = Que::Worker.new(
-  queue: queue_name, wake_interval: wake_interval,
+  queue: queue_name,
+  wake_interval: wake_interval,
   metrics_labels: metrics_labels,
   priority_threshold: priority_threshold,
   lock_cursor_expiry: cursor_expiry
 )
+
 Thread.new { worker.metrics.expose(port: options.metrics_port) } if options.metrics_port
 
-%w[INT TERM].each { |signal| trap(signal) { worker.stop! } }
+stop = false
+%w[INT TERM].each { |signal| trap(signal) { stop = true } }
 
-worker.work_loop
+worker_thread = Thread.new { worker.work_loop }
+
+loop do
+  sleep 0.1
+  break if stop
+end
+
+Que.logger.info(msg: 'Waiting for worker to finish', event: 'que.worker.finish_wait')
+
+worker.stop!
+
+if timeout
+  timeout.seconds.downto(0) do
+    break unless worker_thread.alive?
+    sleep 1
+  end
+
+  if worker_thread.alive?
+    Que.logger.info(msg: 'Worker still running - forcing it to stop', event: 'que.worker.finish_force')
+    worker_thread.raise(QueJobTimeoutError, timeout_message)
```

Comment:
Do you reckon `QueJobTimeoutError` sounds like the job has been running too long in general? I feel like the use of `timeout` in this file isn't obvious, and "signal timeout" or something similar would communicate the intent better in the flags, exception and variables.

Reply:
It kinda does mean the job has been running too long, though? I want the message communicated here to be "this job took too long to complete, so it has been killed".

Reply:
I think that's confusing because "too long" is actually "too long when the worker is being terminated" – it wouldn't necessarily be too long at another timeout. When a developer picks up this exception they'll initially be wondering why this timed out but other jobs are allowed to run for longer, no?

'cause this is only a very specific kind of timeout, I reckon we take away that friction by making it really clear what kind of timeout this is any why it happened!

Reply:
This is going to occur any time a deploy is executed, so in general it's not safe to ever have a `NoRetry` job that runs for longer than this timeout. That's what I think we should be communicating.